File: src/modules/trajectory/services/parsing/AseImportBridge.ts
import { spawn } from 'node:child_process';
import path from 'node:path';
import readline from 'node:readline';
import type {
    ParsedFrameMetadata,
    ParsedSimulationCell
} from '@modules/trajectory/services/parsing/TrajectoryParserFactory';

const BRIDGE_SCRIPT = path.join(__dirname, 'ase_import_bridge.py');

const ASE_PYTHON = process.env['ASE_PYTHON'] ??
    path.join(__dirname, '../../../../../../../.venv-pyatomsk/bin/python');

interface AseFrame {
    frame: number;
    natoms: number;
    cell: [[number, number, number], [number, number, number], [number, number, number]];
    pbc: [boolean, boolean, boolean];
    symbols: string[];
    positions: number[];
    properties: Record<string, number[]>;
    error?: string;
}

const runBridge = (filePath: string): Promise<AseFrame[]> => new Promise((resolve, reject) => {
    const proc = spawn(ASE_PYTHON, [BRIDGE_SCRIPT, filePath], {
        stdio: ['ignore', 'pipe', 'pipe']
    });

    const frames: AseFrame[] = [];
    const rl = readline.createInterface({ input: proc.stdout! });

    rl.on('line', (line) => {
        if (!line.trim()) return;
        try {
            frames.push(JSON.parse(line) as AseFrame);
        } catch {
        }
    });

    proc.stderr?.on('data', (chunk: Buffer) => {
        process.stderr.write(chunk);
    });

    proc.on('error', reject);
    proc.on('close', (code) => {
        if (code !== 0 && frames.length === 0) {
            reject(new Error(`ase_import_bridge exited ${code ?? 'null'} with no output`));
        } else {
            resolve(frames);
        }
    });
});

const cellToSimulationCell = (
    cell: AseFrame['cell'],
    pbc: AseFrame['pbc']
): ParsedSimulationCell => {
    const [a, b, c] = cell;
    const width  = Math.abs(a[0]);
    const length = Math.abs(b[1]);
    const height = Math.abs(c[2]);
    return {
        boundingBox: { width, length, height },
        geometry: {
            cell_vectors: cell,
            cell_origin: [0, 0, 0],
            periodic_boundary_conditions: { x: pbc[0], y: pbc[1], z: pbc[2] }
        }
    };
};

export const parseAseMetadata = async (filePath: string): Promise<ParsedFrameMetadata> => {
    const frames = await runBridge(filePath);
    if (frames.length === 0) {
        throw new Error('Unsupported trajectory format');
    }
    const first = frames[0];
    if (first.error) {
        throw new Error(`ASE bridge: ${first.error}`);
    }
    return {
        timestep: 0,
        natoms: first.natoms,
        headers: ['id', 'type', 'x', 'y', 'z', ...Object.keys(first.properties)],
        simulationCell: cellToSimulationCell(first.cell, first.pbc)
    };
};

File: src/modules/trajectory/services/parsing/ase_import_bridge.py
#!/usr/bin/env python3
"""ASE import bridge — reads a trajectory file via ase.io and streams JSON frames to stdout.

Usage:
    python ase_import_bridge.py <file_path>

Each output line is a JSON object:
  per-frame: { frame: <n>, natoms: <n>, cell: [[3x3]], pbc: [bool,bool,bool],
               symbols: [...], positions: [[x,y,z]...], properties: {name: [...],...} }
  error:     { error: "<message>" }

STDOUT is reserved for the JSON protocol. All diagnostics go to STDERR.
"""
import json
import sys


def _array_to_list(arr):
    try:
        return arr.tolist()
    except AttributeError:
        return list(arr)


def emit(obj):
    sys.stdout.write(json.dumps(obj, separators=(',', ':')) + '\n')
    sys.stdout.flush()


def run(file_path):
    try:
        import ase.io
    except ImportError as exc:
        emit({'error': f'ase not available: {exc}'})
        return

    try:
        frames = ase.io.iread(file_path, index=':')
    except Exception as exc:
        emit({'error': f'ase.io.iread failed: {exc}'})
        return

    for frame_index, atoms in enumerate(frames):
        try:
            cell = _array_to_list(atoms.get_cell())
            pbc = _array_to_list(atoms.get_pbc())
            symbols = list(atoms.get_chemical_symbols())
            positions = _array_to_list(atoms.get_positions())

            properties = {}
            for key, value in atoms.arrays.items():
                if key in ('positions', 'numbers'):
                    continue
                try:
                    properties[key] = _array_to_list(value)
                except Exception:
                    pass

            emit({
                'frame': frame_index,
                'natoms': len(atoms),
                'cell': cell,
                'pbc': pbc,
                'symbols': symbols,
                'positions': positions,
                'properties': properties
            })
        except Exception as exc:
            emit({'error': f'frame {frame_index}: {exc}'})
            return


if __name__ == '__main__':
    if len(sys.argv) < 2:
        emit({'error': 'usage: ase_import_bridge.py <file_path>'})
        sys.exit(1)
    run(sys.argv[1])
